### PR TITLE
UCS/UCT/RCACHE: Add memory usage limits to registration cache

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by yosefe** [comment]
@evgeny-leksikov @brminich can you pls review

**Review by brminich — APPROVED** [review]

File: src/ucs/memory/rcache.c [hunk 1]
```diff
@@ -200,6 +200,72 @@ static void ucs_rcache_find_regions(ucs_rcache_t *rcache, ucs_pgt_addr_t from,
...
+
+/* Lock must be held in write mode */
+static inline void
+ucs_rcache_region_lru_untrack(ucs_rcache_t *rcache, ucs_rcache_region_t *region)
+{
+    --rcache->num_regions;
+    rcache->total_size -= region->super.end - region->super.start;
+
+    ucs_spin_lock(&rcache->lru.lock);
+    ucs_rcache_region_lru_remove(rcache, region);
```
Comment (evgeny-leksikov) [inline src/ucs/memory/rcache.c]: why track/untrack has asymmetry in impl?

File: src/ucs/memory/rcache.c [hunk 2]
```diff
@@ -216,6 +282,7 @@ static void ucs_mem_region_destroy_internal(ucs_rcache_t *rcache,
         }
     }
 
+    ucs_rcache_region_lru_untrack(rcache, region);
```
Comment (evgeny-leksikov) [inline src/ucs/memory/rcache.c]: is it possible to destroy region which was evicted from LRU?

File: test/gtest/ucs/test_rcache.cc [hunk 3]
```diff
@@ -637,6 +647,99 @@ UCS_MT_TEST_F(test_rcache_no_register, merge_invalid_prot_slow, 5)
...
+
+UCS_TEST_F(test_rcache_with_limit, by_size) {
+    static const size_t size = 600;
+
+    /* First region will be added */
+    void *ptr1 = malloc(size);
+    get_put(ptr1, size);
+    EXPECT_EQ(1, m_rcache.get()->num_regions);
+
+    /* Second region will cause removing of first region */
```
Comment (evgeny-leksikov) [inline test/gtest/ucs/test_rcache.cc:714]: maybe add a corner case where 1 elem size > max_size and after adding it the cache should be empty?

File: src/ucs/memory/rcache.c [hunk 4]
```diff
@@ -200,6 +200,72 @@ static void ucs_rcache_find_regions(ucs_rcache_t *rcache, ucs_pgt_addr_t from,
...
+    ++rcache->num_regions;
+    rcache->total_size += region->super.end - region->super.start;
+}
+
+/* Lock must be held in write mode */
+static inline void
+ucs_rcache_region_lru_untrack(ucs_rcache_t *rcache, ucs_rcache_region_t *region)
+{
+    --rcache->num_regions;
+    rcache->total_size -= region->super.end - region->super.start;
```
Comment (evgeny-leksikov) [inline src/ucs/memory/rcache.c]: seems like not LRU related, maybe move it to upper region create/destroy functions?

**Review by evgeny-leksikov — COMMENTED** [review]
IMO `ucs_rcache_region_lru_track`/`ucs_rcache_region_lru_untrack` are redundant and add some confusion

File: src/ucs/memory/rcache.c (hunk 1 above)
Comment (yosefe) [reply src/ucs/memory/rcache.c]: if we added to lru list in "track", it would have been removed immediately after that in "get"

File: src/ucs/memory/rcache.c (hunk 2 above)
Comment (yosefe) [reply src/ucs/memory/rcache.c]: yes, that's why we check "lru_flag" before removing from the list

File: test/gtest/ucs/test_rcache.cc (hunk 3 above)
Comment (yosefe) [reply test/gtest/ucs/test_rcache.cc:714]: used elements are not evicted even if size exceeds usage (otherwise would need to handle registration failure flow)

File: src/ucs/memory/rcache.c (hunk 4 above)
Comment (yosefe) [reply src/ucs/memory/rcache.c]: ok

**Comment by yosefe** [comment]
@evgeny-leksikov fixed

File: test/gtest/ucs/test_rcache.cc (hunk 3 above)
Comment (evgeny-leksikov) [reply test/gtest/ucs/test_rcache.cc:714]: ic, it's evicted from LRU list but used elem is in cache, right?
https://github.com/yosefe/ucx/pull/64/files#diff-90801b9fc899f6d90f9490e6d897c14379956209ee6f20cc289e8d223dc5beb8R448

**Review by evgeny-leksikov — APPROVED** [review]
